virtual_rainforest/core/data.py

```diff
@@ -0,0 +1,362 @@
+"""API documentation for the :mod:`core.data` module.
+**************************************************
+
+This module handles the population and storage of data sources used to run Virtual
+Rainforest simulations.
+
+The Data class
+==============
+
+The core :class:`~virtual_rainforest.core.data.Data` class is a wrapper around an
+:class:`~xarray.DataSet` object, which is used to core variables in the simulation on
+shared axes. The :class:`~xarray.DataSet` object is stored as the :attr:`Data.data`
+attribute, and this attribute can be used directly to access :class:`~xarray.DataSet`
+methods.
+
+The :class:`~virtual_rainforest.core.data.Data` itself provides methods to add data to
+an instance, and to automatically validate added data against the configuration of the
+simulation.
+
+* The :class:`~virtual_rainforest.core.data.Data.__setitem__` method is used to validate
+  input data. The data must be an :class:`~xarray.DataSet` object and it is then checked
+  to see that its dimensions are congruent with the simulation configuration. Valid
+  inputs are then added onto the internal :class:`~xarray.DataSet` instance.
+* The :class:`~virtual_rainforest.core.data.Data.__getitem__` method is used to retrieve
+  a named variable from the internal :class:`~xarray.DataSet` instance.
+* The :class:`~virtual_rainforest.core.data.Data.__contains__` method is provided to
+  test if a named variable is included in the internal :class:`~xarray.DataSet`
+  instance.
+
+..code-block:: python
+
+    grid = Grid()
+    data = Data(grid)
+    data['varname'] = DataArray([1,2,3])
+    'varname' in data
+    data['varname']
+
+Data validation
+---------------
+
+The :meth:`~virtual_rainforest.core.data.Data.__setitem__` method  validates a
+provided :class:`~xarray.DataArray` by passing it to the
+:func:`~virtual_rainforest.core.axes.validate_datarray` function, described in the
+:mod:`~virtual_rainforest.core.axes` module. See that module for discussion of the
+:class:`~virtual_rainforest.core.axes.AxisValidators` class and the concept of core
+axes.
+
+The :class:`~virtual_rainforest.core.data.Data` instance records validation that has
+been applied to each variable on each core axis and provides the
+:class:`~virtual_rainforest.core.data.Data.on_core_axis` method to allow models to
+confirm that required variables have been validated on particular axes.
+
+Adding data from a file
+-----------------------
+
+The general solution for programmatically adding data from a file is to:
+
+* manually open a data file using the appropriate reader packages for the format,
```

Comment:
This isn't rendering as bullet points for some reason

Reply:
I know and it is **maddening**. Some issue with `sphinx` that I got bored of trying to cure for now.

Reply:
Yeah fair, good bug to target for Tuesday's session?

Reply:
Oh god no. Not unless we want to spend hours trying out  combinations of `sphinx` and `sphinx-rtd-theme` versions.  I think something is broken in the implementation...

Reply:
Ahh yeah fair enough